File: 卡码/代码随想录/1/06-二叉树/C++/03-二叉树修改与构造/LeetCode_0105.cpp
//105. 从前序与中序遍历序列构造二叉树
#include <iostream>
#include <vector>
#include "../TreeNode.h"
using namespace std;

class Solution {
public:
	TreeNode* traversal(vector<int>& inOrder,int inLeft,int inRight,vector<int>& preOrder,int preLeft,int preRight){
		if(inLeft>inRight) return nullptr;
		TreeNode* root=new TreeNode(preOrder[preLeft]);
		if(inLeft==inRight) return root;
		int rootIndex=0;
		for(int i=inLeft;i<=inRight;i++){
			if(inOrder[i]==root->val){
				rootIndex=i;
				break;
			}
		}
		int leftLength=rootIndex-inLeft;
		root->left=traversal(inOrder,inLeft,rootIndex-1,preOrder,preLeft+1,preLeft+leftLength);
		root->right=traversal(inOrder,rootIndex+1,inRight,preOrder,preLeft+1+leftLength,preRight);
		return root;
	}
  TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
		if(inorder.size()==0||preorder.size()==0) return nullptr;
		return traversal(inorder,0,inorder.size()-1,preorder,0,preorder.size()-1);
  }
};

File: 卡码/代码随想录/1/06-二叉树/Python/03-二叉树修改与构造/LeetCode_0105.py
#105. 从前序与中序遍历序列构造二叉树
from typing import *
from ..TreeNode import *

class Solution:
  def traversal(self,preOrder:List[int],preLeft:int,preRight:int,inOrder:List[int],inLeft:int,inRight:int)->Optional[TreeNode]:
    if inLeft>inRight:
      return None
    root:TreeNode=TreeNode(preOrder[preLeft],None,None)
    if inLeft==inRight:
      return root
    rootIndex:int=0
    for i in range(inLeft,inRight+1):
      if inOrder[i]==root.val:
        rootIndex=i
        break
    leftLength:int=rootIndex-inLeft
    root.left=self.traversal(preOrder,preLeft+1,preLeft+leftLength,inOrder,inLeft,rootIndex-1)
    root.right=self.traversal(preOrder,preLeft+leftLength+1,preRight,inOrder,rootIndex+1,inRight)
    return root

  def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    if len(inorder)==0 or len(preorder)==0:
      return None
    return self.traversal(preorder,0,len(preorder)-1,inorder,0,len(inorder)-1)
